From: Kevin Locke <kevin@kevinlocke.name>
Date: Mon, 16 Jul 2018 11:37:10 -0600
Subject: Handle asynchronous connection via EINPROGRESS

POSIX/SUS specifies:

    If the connection cannot be established immediately and O_NONBLOCK
    is set for the file descriptor for the socket, connect() shall fail
    and set errno to [EINPROGRESS], but the connection request shall not
    be aborted, and the connection shall be established asynchronously.

This can occur when the request load is higher than Dovecot is
configured to accept or connections are made faster than Dovecot can
keep up.  When this happens, proceed.  The connection will finish
asynchronously in the select loop.

Signed-off-by: Kevin Locke <kevin@kevinlocke.name>
---
 mod_authn_dovecot.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mod_authn_dovecot.c b/mod_authn_dovecot.c
index 84fb7d5..fa69c90 100644
--- a/mod_authn_dovecot.c
+++ b/mod_authn_dovecot.c
@@ -25,6 +25,7 @@
 #include "http_request.h"
 #include "mod_auth.h"
 
+#include <errno.h>
 #include <unistd.h>
 #include <stdio.h>
 #include <string.h>
@@ -170,7 +171,7 @@ static authn_status check_password(request_rec * r, const char *user, const char
 	address.sun_family = AF_UNIX;
 	memcpy(address.sun_path, conf->dovecotauthsocket, authsocklen + 1);
 	result = connect(auths, (struct sockaddr *)&address, sizeof address);
-	if (result) {
+	if (result && errno != EINPROGRESS) {
 		ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r, "Dovecot Authentication: could not connect to dovecot socket %s: %s", address.sun_path, strerror(errno));
 		if (conf->authoritative == 0) {
 			return DECLINED;
